src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,293 @@
+    /// </summary>
+    private readonly TopicMessageHandler messageHandler;
+    /// <summary>
+    /// A reference to the DaprClient instance.
+    /// </summary>
+    private readonly P.Dapr.DaprClient client;
+    /// <summary>
+    /// Flag that prevents the developer from accidentally initializing the subscription more than once from the same receiver.
+    /// </summary>
+    private bool hasInitialized;
+    /// <summary>
+    /// Flag that ensures the instance is only disposed a single time.
+    /// </summary>
+    private bool isDisposed;
+
+    /// <summary>
+    /// Constructs a new instance of a <see cref="PublishSubscribeReceiver"/> instance.
+    /// </summary>
+    /// <param name="pubSubName">The name of the Dapr Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Options allowing the behavior of the receiver to be configured.</param>
+    /// <param name="handler">The delegate reflecting the action to take upon messages received by the subscription.</param>
+    /// <param name="client">A reference to the DaprClient instance.</param>
+    internal PublishSubscribeReceiver(string pubSubName, string topicName, DaprSubscriptionOptions options, TopicMessageHandler handler, P.Dapr.DaprClient client)
+    {
+        this.client = client;
+        this.pubSubName = pubSubName;
+        this.topicName = topicName;
+        this.options = options;
+        this.messageHandler = handler;
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
+    public async Task SubscribeAsync(CancellationToken cancellationToken = default)
+    {
+        //Prevents the receiver from performing the subscribe operation more than once (as the multiple initialization messages would cancel the stream).
+        if (hasInitialized)
+            return;
+        hasInitialized = true;
+
+        var stream = await GetStreamAsync(cancellationToken);
+
+        //Retrieve the messages from the sidecar and write to the messages channel
+        var fetchMessagesTask = FetchDataFromSidecarAsync(stream, topicMessagesChannel.Writer, cancellationToken);
+
+        //Process the messages as they're written to either channel
+        var acknowledgementProcessorTask = ProcessAcknowledgementChannelMessagesAsync(cancellationToken);
+        var topicMessageProcessorTask = ProcessTopicChannelMessagesAsync(cancellationToken);
+
+        try
+        {
+            await Task.WhenAll(fetchMessagesTask, acknowledgementProcessorTask, topicMessageProcessorTask);
+        }
+        catch (OperationCanceledException)
+        {
+            await DisposeAsync();
```

Comment:
Nit: the caller should dispose of the instance regardless of whether it completed naturally, was cancelled, or otherwise failed, so probably no reason to do this here.

Reply:
Good point